charmcraft/commands/store/client.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2020 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""A client to hit the Store."""
+
+import logging
+import os
+import webbrowser
+from http.cookiejar import MozillaCookieJar
+
+import appdirs
+from macaroonbakery import httpbakery
+
+from charmcraft.cmdbase import CommandError
+
+logger = logging.getLogger('charmcraft.commands.store')
+
+# XXX Facundo 2020-06-19: only staging for now; will make it "multi-server" when we have proper
+# functionality in Store's production
+BASE_URL = 'https://api.staging.snapcraft.io/publisher/api'
+
+
+def visit_page_with_browser(visit_url):
+    """Open a browser so the user can validate its identity."""
+    logger.warning(
+        "Opening an authorization web page in your browser; if it does not open, "
+        "please open this URL: %s", visit_url)
+    webbrowser.open(visit_url, new=1)
+
+
+class _AuthHolder:
+    """Holder and wrapper of all authentication bits.
+
+    Main two purposes of this class:
+
+    - deal with credentials persistence
+
+    - wrap HTTP calls to ensure authentication
+
+    XXX Facundo 2020-06-18: right now for functionality bootstrapping we're storing credentials
+    on disk, we may move to a keyring, wallet, other solution, or firmly remain here when we
+    get a "security" recommendation.
+    """
+
+    def __init__(self):
+        self._cookiejar_filepath = appdirs.user_config_dir('charmcraft.credentials')
+        self._cookiejar = None
+        self._client = None
+
+    def clear_credentials(self):
+        """Clear stored credentials."""
+        if os.path.exists(self._cookiejar_filepath):
+            os.unlink(self._cookiejar_filepath)
+            logger.debug("Credentials cleared: file %r removed", self._cookiejar_filepath)
+        else:
+            logger.debug("Credentials file not found to be removed: %r", self._cookiejar_filepath)
+
+    def _save_credentials(self):
```

Comment:
I prefer a name that hints that it might not be done. ensure or _check_save... are helpful to indicate it doesnt always actually perform the action.

Reply:
Renamed to `_save_credentials_if_changed`.